File: internal/sparkline/sparkline.go
package sparkline

import "strings"

// blocks are the Unicode block characters U+2581 through U+2588.
var blocks = []rune{'▁', '▂', '▃', '▄', '▅', '▆', '▇', '█'}

// Render produces a sparkline string of the given width from values.
// Empty input returns spaces of the given width.
func Render(values []float64, width int) string {
	if width <= 0 {
		width = 20
	}
	if len(values) == 0 {
		return strings.Repeat(" ", width)
	}

	resampled := Resample(values, width)

	min, max := resampled[0], resampled[0]
	for _, v := range resampled[1:] {
		if v < min {
			min = v
		}
		if v > max {
			max = v
		}
	}

	runes := make([]rune, len(resampled))
	for i, v := range resampled {
		if max == min {
			runes[i] = blocks[3] // middle block for flat line
		} else {
			norm := (v - min) / (max - min)
			idx := int(norm * 7)
			if idx > 7 {
				idx = 7
			}
			runes[i] = blocks[idx]
		}
	}
	return string(runes)
}

// Resample resamples values to targetLen using linear interpolation.
func Resample(values []float64, targetLen int) []float64 {
	if len(values) == 0 {
		return nil
	}
	if len(values) == targetLen {
		out := make([]float64, targetLen)
		copy(out, values)
		return out
	}

	out := make([]float64, targetLen)
	srcLen := len(values)

	for i := range targetLen {
		var srcIdx float64
		if targetLen == 1 {
			srcIdx = float64(srcLen-1) / 2.0
		} else {
			srcIdx = float64(i) * float64(srcLen-1) / float64(targetLen-1)
		}

		lo := int(srcIdx)
		if lo >= srcLen-1 {
			out[i] = values[srcLen-1]
			continue
		}

		frac := srcIdx - float64(lo)
		out[i] = values[lo]*(1-frac) + values[lo+1]*frac
	}

	return out
}

File: src/utils/sparkline.lua
local M = {}

-- Unicode block characters U+2581 through U+2588
local BLOCKS = {
    "▁", -- U+2581
    "▂", -- U+2582
    "▃", -- U+2583
    "▄", -- U+2584
    "▅", -- U+2585
    "▆", -- U+2586
    "▇", -- U+2587
    "█", -- U+2588
}

function M.render(values, width)
    if not values or #values == 0 then
        return string.rep(" ", width or 20)
    end

    width = width or #values

    -- Resample values to fit width if needed
    local sampled = M.resample(values, width)

    local min_val = math.huge
    local max_val = -math.huge
    for _, v in ipairs(sampled) do
        if v < min_val then min_val = v end
        if v > max_val then max_val = v end
    end

    local range = max_val - min_val
    if range == 0 then
        range = 1
    end

    local result = {}
    for _, v in ipairs(sampled) do
        local normalized = (v - min_val) / range
        local index = math.floor(normalized * 7) + 1
        if index > 8 then index = 8 end
        if index < 1 then index = 1 end
        result[#result + 1] = BLOCKS[index]
    end

    return table.concat(result)
end

function M.resample(values, target_len)
    if #values == target_len then
        return values
    end

    if #values == 0 then
        return {}
    end

    local result = {}
    local ratio = #values / target_len

    for i = 1, target_len do
        local src_idx = (i - 1) * ratio + 1
        local lower = math.floor(src_idx)
        local upper = math.min(lower + 1, #values)
        local frac = src_idx - lower

        if lower == upper or upper > #values then
            result[i] = values[lower]
        else
            result[i] = values[lower] * (1 - frac) + values[upper] * frac
        end
    end

    return result
end

return M
